crates/bevy_ecs/src/schedule/executor/single_threaded.rs

```diff
@@ -44,13 +61,57 @@ impl SystemExecutor for SingleThreadedExecutor {
     }
 
     fn run(&mut self, schedule: &mut SystemSchedule, world: &mut World) {
+        /// return the next system not exempt from stepping starting at, and
+        /// inclusive of, first.  If there are no more systems that can be
+        /// stepped, the first non-exempt system index will be returned.
+        fn next_system(schedule: &SystemSchedule, first: usize) -> usize {
+            for i in first..schedule.systems_with_stepping.len() {
+                if schedule.systems_with_stepping[i] {
+                    return i;
+                }
+            }
+            for i in 0..schedule.systems_with_stepping.len() {
+                if schedule.systems_with_stepping[i] {
+                    return i;
+                }
+            }
+            panic!("all systems exempt from stepping");
+        }
+
+        if self.step_state != StepState::Disabled {
+            match self.step_state {
+                StepState::Disabled => (),
+                StepState::Wait(_) => {
+                    self.completed_systems |= &schedule.systems_with_stepping;
```

Comment:
All of these bitwise or operations require comments: they're tricky to read and reason about.